src/t_list.c

```diff
@@ -394,76 +440,85 @@ void listElementsRemoved(client *c, robj *key, int where, robj *o) {
         dbDelete(c->db, key);
     }
     signalModifiedKey(c, c->db, key);
-    server.dirty++;
+    server.dirty += count;
 }
 
+/* Implements the generic list pop operation for LPOP/RPOP.
+ * The where argument specifies which end of the list is operated on. An
+ * optional count may be provided as the third argument of the client's
+ * command. */
 void popGenericCommand(client *c, int where) {
+    long count = 0;
+    robj *value;
+
+    if (c->argc > 3) {
+        addReplyErrorFormat(c,"wrong number of arguments for '%s' command",
+                            c->cmd->name);
+        return;
+    } else if (c->argc == 3) {
+        /* Parse the optional count argument. */
+        if (getPositiveLongFromObjectOrReply(c,c->argv[2],&count,NULL) != C_OK) 
+            return;
+        if (count == 0) {
+            /* Fast exit path. */
+            addReplyNullArray(c);
+            return;
+        }
+    }
+
     robj *o = lookupKeyWriteOrReply(c, c->argv[1], shared.null[c->resp]);
     if (o == NULL || checkType(c, o, OBJ_LIST))
         return;
 
-    robj *value = listTypePop(o, where);
-    if (value == NULL) {
-        addReplyNull(c);
-    } else {
+    if (!count) {
+        /* Pop a single element. This is POP's original behavior that replies
+        * with a bulk string. */
+        value = listTypePop(o,where);
+        serverAssert(value != NULL);
         addReplyBulk(c,value);
         decrRefCount(value);
-        listElementsRemoved(c,c->argv[1],where,o);
+        listElementsRemoved(c,c->argv[1],where,o,1);
+    } else {
+        /* Pop a range of elements. An addition to the original POP command,
+        *  which replies with a multi-bulk. */
+        int direction = (where == LIST_HEAD) ? LIST_TAIL : LIST_HEAD;
+        long llen = listTypeLength(o);
+        long rangelen = (count > llen) ? llen : count;
+        long rangestart = (direction == LIST_TAIL) ? 0 : -rangelen;
+        long rangeend = (direction == LIST_TAIL)
+                        ? rangelen - 1
+                        : -1;
```

Comment:
why is one ternary in one line and the other in 3? (i vote for 1)
p.s. IMHO referring to `where` in these conditions is less confusing (comparing to `direction`)

Reply:
Resolved via bae0cb8 - switched to where for less confusion and avoided multi line